File: 188__Best_Time_to_Buy_and_Sell_Stock_IV.py
class Solution(object):

    def maxProfit_as_many_transactions(self, prices):
        if prices == []:
            return 0
        result = 0
        for i in range(1, len(prices)):
            if prices[i] > prices[i - 1]:
                result += prices[i] - prices[i - 1]
        return result

    def maxProfit(self, K, prices):
        """
        :type k: int
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        K = min(K, len(prices) - 1)
        if K == 0:
            return 0
        if K >= len(prices) / 2:
            return self.maxProfit_as_many_transactions(prices)

        n = len(prices)
        dp = [[0] * n for _ in range(K + 1)]
        for k in range(1, K + 1):
            tmpMax = -1 * prices[0]
            for i in range(1, n):
                # Do nothing; Buy at temMax, sell at i
                dp[k][i] = max(dp[k][i - 1], prices[i] + tmpMax)
                # Update tempMax: finish k - 1 trades before, and buy at i
                tmpMax = max(tmpMax, dp[k - 1][i - 1] - prices[i])
        return dp[-1][-1]

File: Java/188__Best_Time_to_Buy_and_Sell_Stock_IV.java
class Solution {
    public int maxProfit(int K, int[] prices) {
        if (K == 0 || prices.length == 0) return 0;
        if (K > prices.length / 2) return getMax(prices);
        int N = prices.length;
        int[][] owned = new int[N][K + 1];
        int[][] clear = new int[N][K + 1];

        for (int k = 1; k <= K; k++)
            owned[0][k] = -prices[0];
        for (int i = 1; i < N; i++) {
            for (int k = 1; k <= K; k++) {
                owned[i][k] = Math.max(owned[i - 1][k], clear[i - 1][k - 1] - prices[i]);
                clear[i][k] = Math.max(clear[i - 1][k], owned[i - 1][k] + prices[i]);
            }
        }
        return clear[N - 1][K];
    }

    private int getMax(int[] prices) {
        int result = 0;
        for (int i = 1; i < prices.length; i++) {
            int diff = prices[i] - prices[i - 1];
            result += Math.max(0, diff);
        }
        return result;
    }
}
